### ISSUE TITLE
ZClient memory leaks finalizers on each request

### ISSUE THREAD

**Issue by ntngel1** [zio/zio-http#2956]
**Describe the bug**
We are using ZIO HTTP in production and have encountered an issue with a constantly increasing amount of memory consumed by the service. At some point, the service is unable to free up memory and subsequently restarts.
<img width="842" alt="Screenshot 2024-07-12 at 13 45 44" src="https://github.com/user-attachments/assets/9b943524-55b0-4aec-8f08-ad0745b60a58">

We investigated this issue by taking a memory dump and inspecting it using Eclipse Memory Analyzer. Here is the result:
<img width="849" alt="Screenshot 2024-07-12 at 13 51 04" src="https://github.com/user-attachments/assets/1037f1d0-d3b8-415e-abc2-80cdc832eed5">

We checked the `zio.Scope` class which has a `Running` state with `finalizers: LongMap[Finalizer]`. We set a breakpoint at `zio.Scope:156` within the `addFinalizerExit` method and discovered that each `ZClient` request adds a new finalizer but does not remove it. The longer the application runs, the more finalizers are added.
<img width="1913" alt="Screenshot 2024-07-12 at 13 41 27" src="https://github.com/user-attachments/assets/eb3fd1fa-f82a-4d67-b821-106d33543ea6">

You can download [this memory dump](https://drive.google.com/file/d/17QLbG5--zrjnc7f0Tui_HU_wlYELHOTe) for investigation. 

**To Reproduce**
1. Clone this demo project to reproduce the problem: https://github.com/ntngel1/zio-http-zclient-leak
2. Set a breakpoint at `zio.Scope:156` in the `addFinalizerExit` method.
3. Run the program.
4. Let it run for 10 seconds (it makes 10 requests). Attach a debugger and you will see that there are 10 finalizers added and none removed. The more requests made, the more finalizers are added, leading to increased memory consumption.

**Expected behaviour**
Finalizers should be removed after they are no longer needed to prevent memory leaks.

**Comment by kyri-petrou** [comment]
Hey @ntngel1, thanks for reporting. I think this one is to be blamed on incomplete documentation and suboptimal UX design of the ZClient.

I believe the issue in the reproducer is that the Scope is not being closed after handling of the request, since you're providing a layer that attaches the finalizes to the global Scope.

Can you try this and let me know if the memory leak still exists?

```scala
  val program = ZIO.scoped(for {
    res <- ZClient.request(Request.get(url).addHeaders(headers))
    data <- res.body.asString
    _ <- Console.printLine(data)
  } yield ())
```

In short, you should be wrapping `ZClient.request` calls in `ZIO.scoped`, but only after you've extracted the body from the response.

**Comment by 987Nabil** [comment]
@kyri-petrou what do you think, should we offer a API that we promote as the default, that works more like `acquireReleaseWith`? So you have to handle the body in a callback.

**Comment by ntngel1** [comment]
> Hey @ntngel1, thanks for reporting. I think this one is to be blamed on incomplete documentation and suboptimal UX design of the ZClient.
> 
> I believe the issue in the reproducer is that the Scope is not being closed after handling of the request, since you're providing a layer that attaches the finalizes to the global Scope.
> 
> Can you try this and let me know if the memory leak still exists?
> 
> ```scala
>   val program = ZIO.scoped(for {
>     res <- ZClient.request(Request.get(url).addHeaders(headers))
>     data <- res.body.asString
>     _ <- Console.printLine(data)
>   } yield ())
> ```
> 
> In short, you should be wrapping `ZClient.request` calls in `ZIO.scoped`, but only after you've extracted the body from the response.

Tried it out, and it works 👍 However, it's not immediately obvious that we need to wrap each request using `ZIO.scoped()`. This wasn't mentioned in the documentation, and all examples were written using the global `Scope`. I think it would be better and less error-prone if the library's API took care of this for developers and defined everything explicitly. This way, everyone would use acquireReleaseWith, and there would be less room for error.

**Comment by kyri-petrou** [comment]
> @kyri-petrou what do you think, should we offer a API that we promote as the default, that works more like `acquireReleaseWith`? So you have to handle the body in a callback.

I'm 100% onboard with this 🙏

**Comment by jdegoes** [comment]
/bounty $175 for fix, doc updates, and tests.

**Comment by kyri-petrou** [comment]
/attempt #2956

I'll be working on this alongside #2957

<div id="algora-attempt" />

| [Algora profile](https://console.algora.io/@/kyri-petrou) | Completed bounties | Tech | Active attempts | Options |
| --- | --- | --- | --- | --- |
| @kyri-petrou | &nbsp;&nbsp;&nbsp;23 ZIO bounties <br/> + 2 bounties from 1 project | <div align="center">Scala, Python, <br />Shell & more</div> | <div align="center">﹟2957</div> | [Cancel attempt](https://console.algora.io/api/bounties/clz8gipzk0024m809ty81stai/cancel-attempt) |
